File: src/example_sql.py
"""Show example queries."""

from pathlib import Path
from datetime import datetime
import pandas as pd
import lib.sqlite as db


def queries():
    """Show example queries."""
    cxn = db.connect()
    now = datetime.now()

    report_name = f'sample_queries_{now.strftime("%Y-%m-%d")}.xlsx'
    report_path = Path('output') / report_name

    bbs_sql, bbs_df = _bbs_query(cxn)
    maps_sql, maps_df = _maps_query(cxn)
    ebird_sql, ebird_df = _ebird_query(cxn)
    global_sql, global_df = _global_query(cxn)

    queries = [
        {'dataset': 'Combined', 'query': global_sql},
        {'dataset': 'BBS', 'query': bbs_sql},
        {'dataset': 'MAPS', 'query': maps_sql},
        {'dataset': 'eBird', 'query': ebird_sql}]
    front_df = pd.DataFrame(queries)

    with pd.ExcelWriter(report_path) as writer:
        front_df.to_excel(writer, sheet_name='Queries', index=False)
        global_df.to_excel(writer, sheet_name='Combined', index=False)
        bbs_df.to_excel(writer, sheet_name='BBS', index=False)
        maps_df.to_excel(writer, sheet_name='MAPS', index=False)
        ebird_df.to_excel(writer, sheet_name='eBird', index=False)


def _bbs_query(cxn):
    sql = """
        SELECT *
          FROM dates
          JOIN bbs_events USING (date_id)
          JOIN counts     USING (date_id)
          JOIN bbs_counts USING (count_id)
          JOIN taxons     USING (taxon_id)
         WHERE year BETWEEN 2010 AND 2014
           AND day  BETWEEN  100 AND  200
           AND lng  BETWEEN  -79 AND  -78
           AND lat  BETWEEN   40 AND   44
         LIMIT 100"""
    df = pd.read_sql(sql, cxn)
    df.geopoint = 'BLOB'
    return sql, df


def _maps_query(cxn):
    sql = """
        SELECT *
          FROM dates
          JOIN maps_events USING (date_id)
          JOIN counts      USING (date_id)
          JOIN maps_counts USING (count_id)
          JOIN taxons      USING (taxon_id)
         WHERE year BETWEEN 2010 AND 2014
           AND day  BETWEEN  100 AND  200
           AND lng  BETWEEN  -79 AND  -78
           AND lat  BETWEEN   40 AND   44
         LIMIT 100"""
    df = pd.read_sql(sql, cxn)
    df.geopoint = 'BLOB'
    return sql, df


def _ebird_query(cxn):
    sql = """
        SELECT *
          FROM dates
          JOIN ebird_events USING (date_id)
          JOIN counts       USING (date_id)
          JOIN ebird_counts USING (count_id)
          JOIN taxons       USING (taxon_id)
         WHERE year BETWEEN 2010 AND 2014
           AND day  BETWEEN  100 AND  200
           AND lng  BETWEEN  -79 AND  -78
           AND lat  BETWEEN   40 AND   44
         LIMIT 100"""
    df = pd.read_sql(sql, cxn)
    df.geopoint = 'BLOB'
    return sql, df


def _global_query(cxn):
    sql = """
        SELECT *
          FROM dates
          JOIN counts USING (date_id)
          JOIN taxons USING (taxon_id)
         WHERE year BETWEEN 2010 AND 2014
           AND day  BETWEEN  100 AND  200
           AND lng  BETWEEN  -79 AND  -78
           AND lat  BETWEEN   40 AND   44
         LIMIT 100"""
    df = pd.read_sql(sql, cxn)
    df.geopoint = 'BLOB'
    return sql, df


if __name__ == '__main__':
    queries()

File: R/example_sql.R
library(dplyr)
library(dbplyr)
library("RPostgreSQL")

# cxn <- src_sqlite(path = "data/processed/sightings.sqlite.db", create = FALSE)
pg <- dbDriver("PostgreSQL")
cxn <- dbConnect(pg, user = "username", password = "password",
                 host = "35.221.16.125", port = 5432, dbname = "sightings")

places <- tbl(cxn, "places")
events <- tbl(cxn, "events")
counts <- tbl(cxn, "counts")
taxons <- tbl(cxn, "taxons")


# "dplyr" is lazy and only evaluates as little as possible to create the desired output.
# This helps with efficiency. So, you may want to avoid the as.data.frame() function,
# it is there only to show how to convert the queries into a dataframe.

combined <- places %>%
  inner_join(events, by = "place_id") %>%
  inner_join(counts, by = "event_id") %>%
  inner_join(taxons, by = "taxon_id") %>%
  filter(between(year, 2010, 2014)) %>%
  filter(between(day, 100, 200)) %>%
  filter(between(lng, -73, -72)) %>%
  filter(between(lat, 40, 44)) %>%
  head(100) %>%
  as.data.frame()


bbs <- places %>%
  inner_join(events, by = "place_id") %>%
  inner_join(counts, by = "event_id") %>%
  inner_join(taxons, by = "taxon_id") %>%
  filter(between(year, 2010, 2014)) %>%
  filter(between(day, 100, 200)) %>%
  filter(between(lng, -73, -72)) %>%
  filter(between(lat, 40, 44)) %>%
  head(100) %>%
  as.data.frame()

maps <- places %>%
  inner_join(events, by = "place_id") %>%
  inner_join(counts, by = "event_id") %>%
  inner_join(taxons, by = "taxon_id") %>%
  filter(between(year, 2010, 2014)) %>%
  filter(between(day, 100, 200)) %>%
  filter(between(lng, -73, -72)) %>%
  filter(between(lat, 40, 44)) %>%
  head(100) %>%
  as.data.frame()

ebird <- places %>%
  inner_join(events, by = "place_id") %>%
  inner_join(counts, by = "event_id") %>%
  inner_join(taxons, by = "taxon_id") %>%
  filter(between(year, 2010, 2014)) %>%
  filter(between(day, 100, 200)) %>%
  filter(between(lng, -73, -72)) %>%
  filter(between(lat, 40, 44)) %>%
  head(100) %>%
  as.data.frame()
